src/Interpreters/DatabaseCatalog.cpp

```diff
@@ -936,6 +945,103 @@ void DatabaseCatalog::enqueueDroppedTableCleanup(StorageID table_id, StoragePtr
+            if (table_id.uuid != UUIDHelpers::Nil)
+                continue;
+            if (it->table_id.database_name == table_id.database_name &&
+                it->table_id.table_name == table_id.table_name &&
+                it->drop_time >= latest_drop_time)
+            {
+                latest_drop_time = it->drop_time;
+                it_dropped_table = it;
+                dropped_table = *it;
+            }
+        }
+        if (it_dropped_table == tables_marked_dropped.end())
+            throw Exception(ErrorCodes::UNKNOWN_TABLE,
+                "The drop task of table {} is in progress, has been dropped or the database engine doesn't support it",
+                table_id.getNameForLogs());
+        latest_metadata_dropped_path = it_dropped_table->metadata_path;
+        /// If is_being_undropped of table is true, table drop task will skip this table.
+        it_dropped_table->is_being_undropped = true;
+        String table_metadata_path = getPathForMetadata(it_dropped_table->table_id);
+
+        try
+        {
+            /// a table is successfully marked undropped,
+            /// if and only if its metadata file was moved to a database.
+            renameNoReplace(latest_metadata_dropped_path, table_metadata_path);
+        }
+        catch (...)
+        {
+            it_dropped_table->is_being_undropped = false;
+            throw Exception(
+                ErrorCodes::FS_METADATA_ERROR,
+                "Cannot undrop table {}, failed to rename {} to {}",
+                dropped_table.table_id.getNameForLogs(),
+                latest_metadata_dropped_path,
+                table_metadata_path);
+        }
+
+        tables_marked_dropped.erase(it_dropped_table);
+        [[maybe_unused]] auto removed = tables_marked_dropped_ids.erase(dropped_table.table_id.uuid);
+        assert(removed);
+        CurrentMetrics::sub(CurrentMetrics::TablesToDropQueueSize, 1);
+    }
+
+    LOG_INFO(log, "Trying Undrop table {} from {}", dropped_table.table_id.getNameForLogs(), latest_metadata_dropped_path);
+
+    try
+    {
+        auto wait_dropped_table_not_in_use = [&]()
+        {
+            while (true)
+            {
+                {
+                    std::lock_guard lock(tables_marked_dropped_mutex);
+                    if (dropped_table.table.unique())
+                        return;
+                }
+                std::this_thread::sleep_for(std::chrono::milliseconds(100));
+            }
+        };
+        wait_dropped_table_not_in_use();
+
+        String query = fmt::format("ATTACH TABLE {}", dropped_table.table_id.getFullTableName());
+        auto query_context = Context::createCopy(getContext());
+        /// Attach table needs to acquire ddl guard, that has already been acquired in undrop table,
+        /// and cannot be acquired in the attach table again.
+        query_context->setInDDLGuard(true);
+        executeQuery(query, query_context, true);
```

Comment:
It makes sense to set `force_restore_data` flag, but it's not possible to pass it through `executeQuery`. Sorry, I did not realize that when I suggested using `executeQuery`. We can pass that flag by using `InterpreterCreateQuery` directly:
```
ASTPtr ast_attach = std::make_shared<ASTCreateQuery>();
ast_attach->attach = true;
ast_attach.setDatabase(dropped_table.table_id.database_name);
ast_attach.setTable(dropped_table.table_id.table_name);

InterpreterCreateQuery interpreter(ast, query_context);
interpreter.setForceAttach(true);
interpreter.setForceRestoreData(true);
interpreter.execute();
```

Also it will be possible to add something like `setDontNeedDDLGuard` to `InterpreterCreateQuery` (so it will not pollute `Context` with two more ad-hoc methods)